File: Leetcode/386.go
package Leetcode

func lexicalOrder(n int) []int {
	ans := make([]int, 0)

	for startDigit := 1; startDigit < 10; startDigit++ {
		if len(ans) > n {
			break
		}
		dfsLexicalOrder(startDigit, n, &ans)
	}
	return ans
}

func dfsLexicalOrder(curr, limit int, ans *[]int) {
	if curr > limit {
		return
	}
	*ans = append(*ans, curr)
	if curr*10 > limit {
		return
	}

	for i := range 10 {
		dfsLexicalOrder(curr*10+i, limit, ans)
	}

}

File: Leetcode/386.py
from typing import List


class Solution:
    def lexicalOrder(self, n: int) -> List[int]:
        ans: list[int] = []

        def dfs(curr: int):
            nonlocal n, ans
            if curr <= n:
                ans.append(curr)
            else:
                return
            if curr * 10 > n:
                return
            for i in range(10):
                dfs(curr * 10 + i)

        start_number: int = 1
        while len(ans) < n:
            dfs(start_number)
            start_number += 1
        return ans
